ext/cuda/operators_spectral_element.jl

```diff
@@ -269,14 +268,59 @@ Base.@propagate_inbounds function operator_evaluate(
     Nq = Quadratures.degrees_of_freedom(QS)
     D = Quadratures.differentiation_matrix(FT, QS)
 
-    ∂f∂ξ₁ = D[i, 1] * input[1, j, vt]
-    ∂f∂ξ₂ = D[j, 1] * input[i, 1, vt]
-
-    for k in 2:Nq
-        ∂f∂ξ₁ += D[i, k] * input[k, j, vt]
-        ∂f∂ξ₂ += D[j, k] * input[i, k, vt]
+    if length(input) == 1 # check types
+        (v₁,) = input
+        @inbounds begin
+            ∂f∂ξ₁ = D[i, 1] ⊠ v₁[1, j, vt]
+            ∂f∂ξ₂ = D[j, 1] ⊠ v₁[i, 1, vt]
+            for k in 2:Nq
+                ∂f∂ξ₁ = ∂f∂ξ₁ ⊞ D[i, k] ⊠ v₁[k, j, vt]
+                ∂f∂ξ₂ = ∂f∂ξ₂ ⊞ D[j, k] ⊠ v₁[i, k, vt]
+            end
+        end
+        return Geometry.Covariant12Vector(∂f∂ξ₁, ∂f∂ξ₂)
+    elseif length(input) == 2
+        # Update `shmem`
+        v₁, v₂ = input
+        @inbounds begin
+            ∂f₁∂ξ₁ = D[i, 1] ⊠ v₁[1, j, vt]
+            ∂f₁∂ξ₂ = D[j, 1] ⊠ v₁[i, 1, vt]
+            ∂f₂∂ξ₁ = D[i, 1] ⊠ v₂[1, j, vt]
+            ∂f₂∂ξ₂ = D[j, 1] ⊠ v₂[i, 1, vt]
+            @simd for k in 2:Nq
+                ∂f₁∂ξ₁ = ∂f₁∂ξ₁ ⊞ D[i, k] ⊠ v₁[k, j, vt]
+                ∂f₁∂ξ₂ = ∂f₁∂ξ₂ ⊞ D[j, k] ⊠ v₁[i, k, vt]
+                ∂f₂∂ξ₁ = ∂f₂∂ξ₁ ⊞ D[i, k] ⊠ v₂[k, j, vt]
+                ∂f₂∂ξ₂ = ∂f₂∂ξ₂ ⊞ D[j, k] ⊠ v₂[i, k, vt]
+            end
+        end
+        return Geometry.AxisTensor(
+            (Geometry.Covariant12Axis(), Geometry.UVAxis()),
+            (∂f₁∂ξ₁, ∂f₁∂ξ₂, ∂f₂∂ξ₁, ∂f₂∂ξ₂),
+        )
+    else
+        v₁, v₂, v₃ = input
+        @inbounds begin
+            ∂f₁∂ξ₁ = D[i, 1] ⊠ v₁[1, j, vt]
+            ∂f₁∂ξ₂ = D[j, 1] ⊠ v₁[i, 1, vt]
+            ∂f₂∂ξ₁ = D[i, 1] ⊠ v₂[1, j, vt]
+            ∂f₂∂ξ₂ = D[j, 1] ⊠ v₂[i, 1, vt]
+            ∂f₃∂ξ₁ = D[i, 1] ⊠ v₃[1, j, vt]
+            ∂f₃∂ξ₂ = D[j, 1] ⊠ v₃[i, 1, vt]
+            @simd for k in 2:Nq
+                ∂f₁∂ξ₁ = ∂f₁∂ξ₁ ⊞ D[i, k] ⊠ v₁[k, j, vt]
+                ∂f₁∂ξ₂ = ∂f₁∂ξ₂ ⊞ D[j, k] ⊠ v₁[i, k, vt]
+                ∂f₂∂ξ₁ = ∂f₂∂ξ₁ ⊞ D[i, k] ⊠ v₂[k, j, vt]
+                ∂f₂∂ξ₂ = ∂f₂∂ξ₂ ⊞ D[j, k] ⊠ v₂[i, k, vt]
+                ∂f₃∂ξ₁ = ∂f₃∂ξ₁ ⊞ D[i, k] ⊠ v₃[k, j, vt]
+                ∂f₃∂ξ₂ = ∂f₃∂ξ₂ ⊞ D[j, k] ⊠ v₃[i, k, vt]
+            end
+        end
+        return Geometry.AxisTensor(
+            (Geometry.Covariant12Axis(), Geometry.UVWAxis()),
+            (∂f₁∂ξ₁, ∂f₁∂ξ₂, ∂f₂∂ξ₁, ∂f₂∂ξ₂, ∂f₃∂ξ₁, ∂f₃∂ξ₂),
+        )
```

Comment:
Just to note, I may have done this as

```julia
        e1 = length(input) ≥ 1 ? (mapreduce(k -> D[i, k] ⊠ input[1][k, j, vt], ⊞, 1:Nq), mapreduce(k -> D[j, k] ⊠ input[1][i, k, vt], ⊞, 1:Nq)) : ()
        e2 = length(input) ≥ 2 ? (mapreduce(k -> D[i, k] ⊠ input[2][k, j, vt], ⊞, 1:Nq), mapreduce(k -> D[j, k] ⊠ input[2][i, k, vt], ⊞, 1:Nq)) : ()
        e3 = length(input) ≥ 3 ? (mapreduce(k -> D[i, k] ⊠ input[3][k, j, vt], ⊞, 1:Nq), mapreduce(k -> D[j, k] ⊠ input[3][i, k, vt], ⊞, 1:Nq)) : ()
        t = (e1...,e2...,e3...)

        if length(input) == 1 # check types
            return Geometry.Covariant12Vector(t...)
        elseif length(input) == 2
            return Geometry.AxisTensor((Geometry.Covariant12Axis(), Geometry.UVAxis()), t)
        else
            return Geometry.AxisTensor((Geometry.Covariant12Axis(), Geometry.UVWAxis()), t)
        end
```
but I think the `@simd` may actually be faster (when available), it's also a bit simpler, so I think that what you have is good.

Reply:
Noted, thanks! Perhaps in a future pass we can update these patterns (here and in the `Curl` operators as well).